### PR TITLE
Change snapshot url returned to UUID, migrate older snapshots

### PR DESCRIPTION
**Pull request by mayankkapoor**
Now the url returned after creating a snapshot shows the UUID instead of ID.

### CODE REVIEW

**Comment by sorenisanerd** [comment]
Have you thought about migration paths? If there are existing snapshots, they'll be using the id on the filesystem, so that'll need to change.

**Comment by mayankkapoor** [comment]
Based on a conversation with @sorenh, we'll need to consider migration path for changing to uuid for filesystem paths as well. People who used the V1 API should be able to access snapshots via ID, and people who use newer APIs should be able to access via UUIDs. The recommended approach: 
Add a boolean to some of the models saying whether to expose them in the v1 api. Having a boolean that tracks if a given resource instance should work in the v1 API would help us make that decision. So an old snapshot will have 'available_in_v1_api' Boolean set to true. The newer snapshots won't.

**Comment by anshprat** [comment]
Presently the baggage of old API is very less. As of now, is there any real user/data who is using the older API and needs to be migrated to the new API? Anyone who would crib and complain if they can no longer access old API?

**Comment by sorenisanerd** [comment]
It's a matter of principle and practice. It may cost us an hour or two of effort now, but we'll gain valuable insight into the potential difficulties of dealing with such issues. This will benefit us in the long run.

**Comment by sorenisanerd** [comment]
As discussed elsewhere, the plan is:

1) Add the visible_in_v1_api field to the snapshot model. Make it default to False, but set it to True for all current snapshots. (The easiest is probably to set default=True, run makemigrations, change it to False, run makemigrations again).
2) Make the v1 ViewSet's get_queryset filter out all snapshots that have visible_in_v1_api = False.
3) Rename all the directories on disk from <id> format to <uuid> format.
4) For all resources that have visible_in_v1_api True, provide a symlink from <id> to <uuid>.

Step 3 and 4 will be done using Django's migration framework. It's meant for database migrations, but we see no reason why we couldn't use it for this purpose as well (using RunPython migrations).

**Comment by mayankkapoor** [comment]
@sorenh 
1. Didn't see your comments before implementing, so step 1 is different. Trying to set to True in migration itself. Let me know if this isn't recommended.
2. Somehow the pypy build with django<1.9 is failing due to "MySQL server has gone away". Suspect this issue is similar to the Travis CI MySQL issue from before.

**Comment by sorenisanerd** [comment]
When you add a new field, all existing rows in the database get the default value assigned to them. That should be True (principle of least surprise: We don't want all snapshots to suddenly vanish from the v1 API (and their ID based URL's should also keep working)). However, going forward, we want the default to be False, except if the snapshot is created through the v1 API. Changing the default in a second migration doesn't change the existing data in the database. It just says what will be the default going forward.

I don't know for sure, but I have a hunch that the pypy thing is because of a leaked connection to MySQL:

In "regular" Python (CPython), file pointers are garbage collected once their last reference disappears. Like, say, if you have a method that opens a file, does some stuff, and then returns, CPython will close the file (because there are no references to the file left). With pypy, it works differently. If you don't explicitly close the file descriptor, it'll stay open until pypy happens to run its garbage collector and finds the unreferenced file objects and destroys it. That's why I always use:

```
with open(...) as fp:
    do_stuff_with_fp
do_stuff_that_doesn't_need_the_fp_anymore
```

When that "with" block terminates, the fp is closed. The fp object might not get garbage collected right then, but at least the file descriptor is closed.

I'm guessing this is the same problem, except with a connection to MySQL. I don't know, though. For now, I'll be moving pypy to "allow_failures" on Travis. I want to make sure it works (I want to use pypy eventually), but I also want to keep being able to trust the output from Travis.

**Comment by mayankkapoor** [comment]
@sorenh ready for review. Made the changes suggested.

File: aasemble/django/apps/api/v1/serializers.py [hunk 1]
```diff
@@ -109,10 +109,14 @@ def create(self, validated_data):
                 if 'tags' in validated_data:
                     tags_data = validated_data.pop('tags')
                     snapshot = mirrorsvc_models.Snapshot.objects.create(**validated_data)
```
Comment (sorenisanerd) [inline aasemble/django/apps/api/v1/serializers.py]: You could change the create call to: `create(visible_in_v1=(selff.view_prefix == 'v1'), **validated_data)`

File: aasemble/django/apps/api/v1/views.py [hunk 2]
```diff
@@ -108,7 +108,7 @@ def get_queryset(self):
                 if self.request.user.is_superuser:
                     qs = self.queryset.all()
                 else:
-                    qs = self.queryset.filter(mirrorset__owner_id=self.request.user.id)
+                    qs = self.queryset.filter(mirrorset__owner_id=self.request.user.id).exclude(visible_to_v1_api=False)
```
Comment (sorenisanerd) [inline aasemble/django/apps/api/v1/views.py:115]: Instead, just add a new conditional:

```
if selff.view_prefix = 'v1':
    qs = qs.exclude(visible_to_v1_api=False)
```

File: aasemble/django/apps/api/v2/views.py [hunk 3]
```diff
@@ -29,3 +32,32 @@ def BuildViewSetFactory(selff):
...
+                    qs.filter(tags__tag=tag)
+
+                return qs
+
+            def perform_update(self, serializer):
+                if 'mirrorset' in self.request.data or 'timestamp' in self.request.data:
+                    raise ValidationError({'detail': 'Method "PATCH" not allowed.'})
+                serializer.save(owner=self.request.user)
+
+        return SnapshotViewSet
```
Comment (sorenisanerd) [inline aasemble/django/apps/api/v2/views.py]: If you make the changes I suggested, you don't need any of this.

**Comment by sorenisanerd** [comment]
We need tests to verify that snapshots created through v1 get `visible_in_v1_api=True` and snapshots created through any other version has it set to False. We also need to verify that they are correctly filtered.

**Comment by mayankkapoor** [comment]
@sorenh incorporated feedback, created tests. In the tests, I faced issues with asserting the timestamps for snapshots are equal to response. So only asserting few elements are as expected.

File: aasemble/django/apps/api/v1/serializers.py [hunk 4]
```diff
@@ -125,7 +125,7 @@ def update(self, instance, validated_data):
 
             class Meta:
                 model = mirrorsvc_models.Snapshot
-                fields = ('self', 'timestamp', 'mirrorset')
+                fields = ('self', 'timestamp', 'mirrorset', 'visible_to_v1_api')
```
Comment (sorenisanerd) [inline aasemble/django/apps/api/v1/serializers.py]: Sorry, I must have missed this when I reviewed it earlier. Why should this be part of the public API? I was thinking of it as an internal bit of book keeping.

**Comment by sorenisanerd** [comment]
I'm sorry I'm dragging my feet on this one. I guess I'm a bit concerned, since it makes changes to the filesystem which we haven't done before. Let's get the last wrinkle worked out, and get it merged, shall we?

**Comment by mayankkapoor** [comment]
Changes:
- removed visible_to_v1_flag from being seen in API
- Updated fixture timestamps, removed milliseconds for v1 snapshots
- Updated tests

**Comment by mayankkapoor** [comment]
```
FAIL: test_snapshot_create_sets_visible_flag_properly (aasemble.django.apps.api.tests.APIv1Tests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/aaSemble/python-aasemble.django/aasemble/django/apps/api/tests.py", line 977, in test_snapshot_create_sets_visible_flag_properly
    self.assertEqual(snapshot_object.visible_to_v1_api, True)
AssertionError: False != True
```

@sorenh Need some advice. I'm observing some different behavior on MySQL (Travis) as opposed to SQLite. This test passes on SQLite tests. Somehow on Travis, if a new snapshot gets created using V1 API, it still gets the visible_to_v1_api flag set to False. We explicity set it to true in v1/serializers.py>SnapshotSerializer>create

```
snapshot = mirrorsvc_models.Snapshot.objects.create(visible_to_v1_api=(selff.view_prefix == 'v1'), **validated_data)
```

**Comment by sorenisanerd** [comment]
I think the problem is something other than you think.

You create the snapshot, then attempt to get its ID by using obj['self'][-2:-1]. If e.g. obj['self'] = `/something/63/`, then you'll get id = 3, not 63.

I'm not sure about the details, but with sqlite, I see the assigned ID be 53. With MySQL, it's 68. At least it was for me. Your mileage may vary. In either case, you're not getting the snapshot you just created.

You can either something like obj['self'].split('/')[-2] to get the whole ID.

**Comment by mayankkapoor** [comment]
Of course. I just thought of it as well. Sorry for the bother. Now I feel stupid.

**Comment by sorenisanerd** [comment]
Np. Happens to the best of us.

**Comment by sorenisanerd** [comment]
@mayankkapoor I'm happy to report that the migration seems to have worked as expected. Great work!

**Comment by mayankkapoor** [comment]
@sorenh Awesome. Thanks for all the guidance.

**Pull request merged**